File: 394-decode-string/decode-string.java

class Solution {
    public String decodeString(String s) {
        StringBuilder result = new StringBuilder(); 
        result.append(s); 
        StringBuilder subString = new StringBuilder();  
        int k = 0;  
        Stack<Integer> countStack = new Stack<>();
        Stack<StringBuilder> stringStack = new Stack<>();

        for (int i = 0; i < s.length(); i++) {

            if (Character.isDigit(s.charAt(i))){
                k = k*10 + (s.charAt(i) - '0');
            }
            else if (s.charAt(i) == '['){
                countStack.push(k);
                stringStack.push(subString);
                subString = new StringBuilder();
                k = 0;
            }
            else if (s.charAt(i) == ']'){
                int repeat = countStack.pop();
                result = stringStack.pop();
                for (int j = 0; j < repeat; j++) {
                    result.append(subString.toString());
                }
                subString = result;
                if (!countStack.isEmpty()){
                    result = new StringBuilder();

                }
            }
            else {
                subString.append(s.charAt(i));
            }
        }

        return result.toString();
}
}

File: 394-decode-string/decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:

        # stack = []
        # result = ""
        # digit = 0
        # for char in s:
        #     if char.isdigit():
        #         if result != "":
        #             stack.append(result)
        #             result=""
        #         digit = digit*10 + int(char)
        #         continue
        #     if char == "[":
        #         stack.append(digit)
        #         digit = 0
        #         continue
        #     if char.isalpha():
        #         result += char
        #         continue
        #     if char == "]":
        #         popchar = stack.pop()
        #         if isinstance(popchar,int):
        #             temp =""
        #             for i in range(int(popchar)):
        #                 temp += result
        #             result = temp
        #             continue
        #         else:
        #             temp = popchar + result
        #             result = temp
        #             continue
        
        # while len(stack) != 0:
        #     popchar = stack.pop()
        #     if isinstance(popchar,int):
        #             temp =""
        #             for i in range(int(popchar)):
        #                 temp += result
        #             result = temp
        #             continue
        #     else:
        #         temp = popchar + result
        #         result = temp
        #         continue
        # return result

        stack = []
        current_string = ""
        current_number = 0

        for char in s:
            if char.isdigit():
                current_number = current_number * 10 + int(char)
            elif char == "[":
                stack.append((current_string, current_number))
                current_string = ""
                current_number = 0
            elif char == "]":
                prev_string, number = stack.pop()
                current_string = prev_string + current_string * number
            else:
                current_string += char

        return current_string


